File: leetcode/5-symmetric-tree/myHan.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def _isSymmetric(self, l, r):
        if l is None and r is None:
            return True
        elif l is None:
            return False
        elif r is None:
            return False
        elif l.val != r.val:
            return False
        else:
            return self._isSymmetric(l.left, r.right) and self._isSymmetric(l.right, r.left)
    def isSymmetric(self, root: TreeNode) -> bool:
        return self._isSymmetric(root.left, root.right)

File: leetcode/9-binary-tree-cameras/myHan.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private static int solve(TreeNode root, HashSet<Integer> isCovered, HashSet<Integer> isCamera) {
        if(root == null) return 0;
        int ret = solve(root.left, isCovered, isCamera) + solve(root.right, isCovered, isCamera);
        if(
            (root.left != null && isCamera.contains(root.left.hashCode()))
            || (root.right != null && isCamera.contains(root.right.hashCode()))
        ) {
            isCovered.add(root.hashCode());
        }
        if(
            (root.left != null && !isCovered.contains(root.left.hashCode()))
            || (root.right != null && !isCovered.contains(root.right.hashCode()))
        ) {
            isCamera.add(root.hashCode());
            isCovered.add(root.hashCode());
            if(root.left != null) isCovered.add(root.left.hashCode());
            if(root.right != null) isCovered.add(root.right.hashCode());
            ret++;
        }
        
        return ret;
    }
    public int minCameraCover(TreeNode root) {
        HashSet<Integer> isCovered = new HashSet<Integer>();
        HashSet<Integer> isCamera = new HashSet<Integer>();
        int ans = solve(root, isCovered, isCamera);
        if(!isCovered.contains(root.hashCode())) ans++;
        return ans;
    }
}